/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

syntax = "proto3";

package experimental;

import "google/protobuf/descriptor.proto";

message AliasingFieldPair {
  google.protobuf.FieldDescriptorProto first = 1;
  google.protobuf.FieldDescriptorProto second = 2;
  enum Direction {
    FORWARD = 0;
    BACKWARD = 1;
    BIDRECTIONAL = 2;
  }
  Direction direction = 3;
}

message Service {
  string name = 1;
  message Method {
    string name = 1;
    google.protobuf.DescriptorProto req = 2;
    google.protobuf.DescriptorProto resp = 3;
  }
  repeated Method methods = 2;
}

// Lists the aliasing fields of the request & response messages of two methods from a service.
message AliasingMethodPair {
  Service.Method method1 = 1;
  Service.Method method2 = 2;
  repeated AliasingFieldPair reqs = 3;
  repeated AliasingFieldPair resps = 4;
  double p = 5;
}
